src/function.ts

```diff
@@ -33,17 +35,34 @@ export class Function<P, O> {
   readonly __functionBrand: ConditionalFunction<P, O>;
 
   constructor(readonly resource: aws_lambda.IFunction) {
-    return Object.assign(lambda, this);
-
-    function lambda(call: CallExpr, vtl: VTL): string {
+    return makeCallable(this, (call: CallExpr, context: VTL | ASL): any => {
       const payloadArg = call.getArgument("payload");
-      const payload = payloadArg ? vtl.eval(payloadArg.expr) : "$null";
 
-      const request = vtl.var(
-        `{"version": "2018-05-29", "operation": "Invoke", "payload": ${payload}}`
-      );
-      return vtl.json(request);
-    }
+      if (isVTL(context)) {
+        const payload = payloadArg ? context.eval(payloadArg.expr) : "$null";
+
+        const request = context.var(
+          `{"version": "2018-05-29", "operation": "Invoke", "payload": ${payload}}`
+        );
+        return context.json(request);
+      } else if (isASL(context)) {
+        this.resource.grantInvoke(context.role);
+        const task: Partial<Task> = {
+          Type: "Task",
+          Resource: "arn:aws:states:::lambda:invoke",
+          Parameters: {
+            FunctionName: this.resource.functionName,
+            [`Payload${
+              payloadArg && isVariableReference(payloadArg.expr) ? ".$" : ""
+            }`]: payloadArg ? ASL.toJson(payloadArg.expr) : null,
+          },
+        };
+        return task;
+      } else {
```

Comment:
I still think the integration code could be handled by the caller, not the callee. This will eventally be a large mix of different languages for representing a lambda call. Just throwing it out there again :-)

Reply:
Hopefully we can keep it clean then - maybe we can generalize the interface of `context` instead of adhoc implementations in the interpreter. The integration logic belongs here IMO, e.g. `grantInvoke`, `grantRead`,